### ISSUE TITLE
[Bug?]: Prerender with dynamic route renders incorrect `<MetaTags>` values

### ISSUE THREAD

**Issue by ched-dev** [redwoodjs/redwood#6981]
### What's not working?

When I have a dynamic route with a `.routeHooks.ts` file, the values of `<MetaTags>` will always render the last item returned from the route hooks file.

### How do we reproduce the bug?

- Start with a fresh install of Redwood JS 3.5.0
- Generate a new page with `yarn rw g page PostPage`
- Convert to dynamic route with `prerender` attribute (`<Route path="/post/{id}" page={PostPage} name="post" prerender />`)
- Add a `PostPage.routeHooks.ts` sibling file with demo data ([example source](https://github.com/ched-dev/redwood-prerender-bug/commit/e1cbcf2cf7152bd30b031ae2406f4155485fe334#diff-27e5a241458cd07cdce34c13be58e6d0f2f16474e20b6d79eb4e6fb7b18d3abb))
- Inside the `PostPage` use data from new route hooks file ([example source](https://github.com/ched-dev/redwood-prerender-bug/commit/e1cbcf2cf7152bd30b031ae2406f4155485fe334#diff-329ccb28c60ca1aba86759cb1a398196626abb34214a004410bbfdf1f7fd5b35))
- Run `yarn rw build`
- Inspect `web/dist/post/*.html` output and see it printed out the same meta tag values for all pages, always using the last value from `PostPage.routeHooks.ts`. Page contents use the correct values though. ([example output](https://github.com/ched-dev/redwood-prerender-bug/commit/e1cbcf2cf7152bd30b031ae2406f4155485fe334#diff-4a85f705371f7e0bd94fba21cea633a024d321781b969f346b4621449828b813))

Full commit diff of my changes: https://github.com/ched-dev/redwood-prerender-bug/commit/e1cbcf2cf7152bd30b031ae2406f4155485fe334

### What's your environment? (If it applies)

```shell
System:
    OS: macOS 12.0.1
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.2 - /private/var/folders/j_/xzmcnyjs0pl630cpk7cd3bc00000gn/T/xfs-113c7611/node
    Yarn: 3.2.3 - /private/var/folders/j_/xzmcnyjs0pl630cpk7cd3bc00000gn/T/xfs-113c7611/yarn
  Databases:
    SQLite: 3.36.0 - /usr/bin/sqlite3
  Browsers:
    Chrome: 107.0.5304.110
    Firefox: 107.0
    Safari: 15.1
  npmPackages:
    @redwoodjs/core: 3.5.0 => 3.5.0
```

### Are you interested in working on this?

- [X] I'm interested in working on this

**Comment by bighitbiker3** [comment]
@ched-dev Were you able to make any progress or workaround this? Currently running into the same thing :/

**Comment by bighitbiker3** [comment]
@ched-dev unfortunately I need the pre-rendered meta tags for share links :/.  My approach is hacky, but it works. I've got this file in my scripts folder which is my modified `prerenderHandler.js` file:

```'use strict'

var _Object$defineProperty = require('@babel/runtime-corejs3/core-js/object/define-property')
var _interopRequireDefault =
  require('@babel/runtime-corejs3/helpers/interopRequireDefault').default
_Object$defineProperty(exports, '__esModule', {
  value: true,
})
exports.handler = exports.getTasks = void 0
var _map = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/map')
)
var _forEach = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/for-each')
)
var _entries = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/object/entries')
)
var _filter = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/filter')
)
var _promise = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/promise')
)
var _flatMap = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/flat-map')
)
var _flat = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/flat')
)
var _repeat = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/repeat')
)
var _some = _interopRequireDefault(
  require('@babel/runtime-corejs3/core-js/instance/some')
)
var _fs = _interopRequireDefault(require('fs'))
var _path = _interopRequireDefault(require('path'))
var _listr = require('listr2')
var _paths = require('@redwoodjs/internal/dist/paths')
var _prerender = require('@redwoodjs/prerender')
var _detection = require('@redwoodjs/prerender/detection')
var _telemetry = require('@redwoodjs/telemetry')
var _colors = _interopRequireDefault(require('../lib/colors'))
var _exec = require('../lib/exec')
class PathParamError extends Error {}
const mapRouterPathToHtml = (routerPath) => {
  if (routerPath === '/') {
    return 'web/dist/index.html'
  } else {
    return `web/dist${routerPath}.html`
  }
}
function getRouteHooksFilePath(routeFilePath) {
  const routeHooksFilePathTs = routeFilePath.replace(
    /\.(js|tsx)$/,
    '.routeHooks.ts'
  )
  if (_fs.default.existsSync(routeHooksFilePathTs)) {
    return routeHooksFilePathTs
  }
  const routeHooksFilePathJs = routeFilePath.replace(
    /\.(js|tsx)$/,
    '.routeHooks.js'
  )
  if (_fs.default.existsSync(routeHooksFilePathJs)) {
    return routeHooksFilePathJs
  }
  return undefined
}

/**
 * Takes a route with a path like /blog-post/{id:Int}
 * Reads path parameters from BlogPostPage.routeHooks.js and returns a list of
 * routes with the path parameter placeholders (like {id:Int}) replaced by
 * actual values
 *
 * So for values like [{ id: 1 }, { id: 2 }, { id: 3 }] (and, again, a route
 * path like /blog-post/{id:Int}) it will return three routes with the paths
 * /blog-post/1
 * /blog-post/2
 * /blog-post/3
 *
 * The paths will be strings. Parsing those path parameters to the correct
 * datatype according to the type notation ("Int" in the example above) will
 * be handled by the normal router functions, just like when rendering in a
 * client browser
 *
 * Example `route` parameter
 * {
 *   name: 'blogPost',
 *   path: '/blog-post/{id:Int}',
 *   routePath: '/blog-post/{id:Int}',
 *   hasParams: true,
 *   id: 'file:///Users/tobbe/tmp/rw-prerender-cell-ts/web/src/Routes.tsx 1959',
 *   isNotFound: false,
 *   filePath: '/Users/tobbe/tmp/rw-prerender-cell-ts/web/src/pages/BlogPostPage/BlogPostPage.tsx'
 * }
 *
 * When returning from this function, `path` in the above example will have
 * been replaced by an actual url, like /blog-post/15
 */
async function expandRouteParameters(route) {
  const routeHooksFilePath = getRouteHooksFilePath(route.filePath)
  if (!routeHooksFilePath) {
    return [route]
  }
  try {
    const routeParameters = await (0, _exec.runScriptFunction)({
      path: routeHooksFilePath,
      functionName: 'routeParameters',
      args: {
        name: route.name,
        path: route.path,
        routePath: route.routePath,
        filePath: route.filePath,
      },
    })
    if (routeParameters) {
      return (0, _map.default)(routeParameters).call(
        routeParameters,
        (pathParamValues) => {
          var _context
          let newPath = route.path
          ;(0, _forEach.default)(
            (_context = (0, _entries.default)(pathParamValues))
          ).call(_context, ([paramName, paramValue]) => {
            newPath = newPath.replace(
              new RegExp(`{${paramName}:?[^}]*}`),
              paramValue
            )
          })
          return {
            ...route,
            path: newPath,
          }
        }
      )
    }
  } catch (e) {
    console.error(_colors.default.error(e.stack))
    return [route]
  }
  return [route]
}

// This is used directly in build.js for nested ListrTasks
const getTasks = async (dryrun, routerPathFilter = null) => {
  var _context2, _context3
  const prerenderRoutes = (0, _filter.default)(
    (_context2 = (0, _detection.detectPrerenderRoutes)())
  ).call(_context2, (route) => route.path)
  const indexHtmlPath = _path.default.join(
    (0, _paths.getPaths)().web.dist,
    'index.html'
  )
  if (prerenderRoutes.length === 0) {
    console.log('\nSkipping prerender...')
    console.log(
      _colors.default.warning(
        'You have not marked any routes with a path as `prerender` in `Routes.{js,tsx}` \n'
      )
    )

    // Don't error out
    return []
  }
  if (!_fs.default.existsSync(indexHtmlPath)) {
    console.error(
      'You must run `yarn rw build web` before trying to prerender.'
    )
    process.exit(1)
    // TODO: Run this automatically at this point.
  }

  ;(0, _exec.configureBabel)()
  const expandedRouteParameters = await _promise.default.all(
    (0, _map.default)(prerenderRoutes).call(prerenderRoutes, (route) =>
      expandRouteParameters(route)
    )
  )

  // Elliott hack to get around

  for (const route of expandedRouteParameters.flat(2)) {
    console.log('Doing', route)
    const outputHtmlPath = mapRouterPathToHtml(route.path)
    const queryCache = {}
    // Check if route param templates in e.g. /path/{param1} have been replaced
    if (/\{.*}/.test(route.path)) {
      throw new PathParamError(
        `Could not retrieve route parameters for ${route.path}`
      )
    }
    try {
      const prerenderedHtml = await (0, _prerender.runPrerender)({
        queryCache,
        renderPath: route.path,
      })
      if (!dryrun) {
        ;(0, _prerender.writePrerenderedHtmlFile)(
          outputHtmlPath,
          prerenderedHtml
        )
      }
    } catch (e) {
      var _context4, _context5
      console.log()
      console.log(
        _colors.default.warning(
          'You can use `yarn rw prerender --dry-run` to debug'
        )
      )
      console.log()
      console.log(
        `${_colors.default.info(
          (0, _repeat.default)((_context4 = '-')).call(_context4, 10)
        )} Error rendering path "${route.path}" ${_colors.default.info(
          (0, _repeat.default)((_context5 = '-')).call(_context5, 10)
        )}`
      )
      ;(0, _telemetry.errorTelemetry)(
        process.argv,
        `Error prerendering: ${e.message}`
      )
      console.error(_colors.default.error(e.stack))
      console.log()
      throw new Error(`Failed to render "${route.filePath}"`)
    }
  }

  const listrTasks = (0, _flatMap.default)(
    (_context3 = (0, _flat.default)(expandedRouteParameters).call(
      expandedRouteParameters
    ))
  ).call(_context3, (routeToPrerender) => {
    // Filter out routes that don't match the supplied routePathFilter
    if (routerPathFilter && routeToPrerender.path !== routerPathFilter) {
      return []
    }
    const outputHtmlPath = mapRouterPathToHtml(routeToPrerender.path)

    // queryCache will be filled with the queries from all the Cells we
    // encounter while prerendering, and the result from executing those
    // queries.
    // We have this cache here because we can potentially reuse result data
    // between different pages. I.e. if the same query, with the same
    // variables is encountered twice, we'd only have to execute it once and
    // then just reuse the cached result the second time.
    return [
      {
        title: `Prerendering boss ${routeToPrerender.path} -> ${outputHtmlPath}`,
        task: async () => {
          return console.log('CHECK THE GH ISSUE - https://github.com/redwoodjs/redwood/issues/6981#issuecomment-1357009132')
          const queryCache = {}
          // Check if route param templates in e.g. /path/{param1} have been replaced
          if (/\{.*}/.test(routeToPrerender.path)) {
            throw new PathParamError(
              `Could not retrieve route parameters for ${routeToPrerender.path}`
            )
          }
          try {
            const prerenderedHtml = await (0, _prerender.runPrerender)({
              queryCache,
              renderPath: routeToPrerender.path,
            })
            if (!dryrun) {
              ;(0, _prerender.writePrerenderedHtmlFile)(
                outputHtmlPath,
                prerenderedHtml
              )
            }
          } catch (e) {
            var _context4, _context5
            console.log()
            console.log(
              _colors.default.warning(
                'You can use `yarn rw prerender --dry-run` to debug'
              )
            )
            console.log()
            console.log(
              `${_colors.default.info(
                (0, _repeat.default)((_context4 = '-')).call(_context4, 10)
              )} Error rendering path "${
                routeToPrerender.path
              }" ${_colors.default.info(
                (0, _repeat.default)((_context5 = '-')).call(_context5, 10)
              )}`
            )
            ;(0, _telemetry.errorTelemetry)(
              process.argv,
              `Error prerendering: ${e.message}`
            )
            console.error(_colors.default.error(e.stack))
            console.log()
            throw new Error(`Failed to render "${routeToPrerender.filePath}"`)
          }
        },
      },
    ]
  })

  return listrTasks
}
exports.getTasks = getTasks
const diagnosticCheck = () => {
  const checks = [
    {
      message: 'Duplicate React version found in web/node_modules',
      failure: _fs.default.existsSync(
        _path.default.join(
          (0, _paths.getPaths)().web.base,
          'node_modules/react'
        )
      ),
    },
    {
      message: 'Duplicate react-dom version found in web/node_modules',
      failure: _fs.default.existsSync(
        _path.default.join(
          (0, _paths.getPaths)().web.base,
          'node_modules/react-dom'
        )
      ),
    },
    {
      message: 'Duplicate core-js version found in web/node_modules',
      failure: _fs.default.existsSync(
        _path.default.join(
          (0, _paths.getPaths)().web.base,
          'node_modules/core-js'
        )
      ),
    },
    {
      message: 'Duplicate @redwoodjs/web version found in web/node_modules',
      failure: _fs.default.existsSync(
        _path.default.join(
          (0, _paths.getPaths)().web.base,
          'node_modules/@redwoodjs/web'
        )
      ),
    },
  ]
  console.log('Running diagnostic checks')
  if ((0, _some.default)(checks).call(checks, (checks) => checks.failure)) {
    var _context6, _context7, _context8
    console.error(
      _colors.default.error('node_modules are being duplicated in `./web` \n')
    )
    console.log('⚠️  Issues found: ')
    console.log((0, _repeat.default)((_context6 = '-')).call(_context6, 50))
    ;(0, _forEach.default)(
      (_context7 = (0, _filter.default)(checks).call(
        checks,
        (check) => check.failure
      ))
    ).call(_context7, (check, i) => {
      console.log(`${i + 1}. ${check.message}`)
    })
    console.log((0, _repeat.default)((_context8 = '-')).call(_context8, 50))
    console.log(
      'Diagnostic check found issues. See the Redwood Forum link below for help:'
    )
    console.log(
      _colors.default.underline(
        'https://community.redwoodjs.com/search?q=duplicate%20package%20found'
      )
    )
    console.log()

    // Exit, no need to show other messages
    process.exit(1)
  } else {
    console.log('✔ Diagnostics checks passed \n')
  }
}
const handler = async ({ path: routerPath, dryRun, verbose }) => {
  const listrTasks = await getTasks(dryRun, routerPath)
  const tasks = new _listr.Listr(listrTasks, {
    renderer: verbose ? 'verbose' : 'default',
    rendererOptions: {
      collapse: false,
    },
    concurrent: !dryRun,
  })
  try {
    if (dryRun) {
      console.log(_colors.default.info('::: Dry run, not writing changes :::'))
    }
    await tasks.run()
  } catch (e) {
    console.log()
    await diagnosticCheck()
    console.log(_colors.default.warning('Tips:'))
    if (e instanceof PathParamError) {
      console.log(
        _colors.default.info(
          "- You most likely need to add or update a *.routeHooks.{js,ts} file next to the Page you're trying to prerender"
        )
      )
    } else {
      console.log(
        _colors.default.info(
          `- This could mean that a library you're using does not support SSR.`
        )
      )
      console.log(
        _colors.default.info(
          '- Avoid using `window` in the initial render path through your React components without checks. \n  See https://redwoodjs.com/docs/prerender#prerender-utils'
        )
      )
      console.log(
        _colors.default.info(
          '- Avoid prerendering Cells with authenticated queries, by conditionally rendering them.\n  See https://redwoodjs.com/docs/prerender#common-warnings--errors'
        )
      )
    }
    console.log()
    process.exit(1)
  }
}
exports.handler = handler
```

And then am running a post install script that looks like 

```import fs from 'fs'

const preRenderFix = () => {
  const fileData = fs.readFileSync('scripts/data/prerenderHandler.js')
  fs.writeFileSync(
    'node_modules/@redwoodjs/cli/dist/commands/prerenderHandler.js',
    fileData,
    { encoding: 'utf-8', flag: 'w' }
  )
}
const execute = async () => {
  console.log('executing post install')
  preRenderFix()
}

execute()
```

Ugly, but it works. I couldn't get `patch-package` to work in this repo for some reason so resorted to this.

**Comment by dac09** [comment]
@bighitbiker3 could you elaborate what the difference is in your script please? 

I'll look to fix this!

**Comment by bighitbiker3** [comment]
@dac09 it's hard to read without syntax highlighting! I'll try and explain better. 

In [this](https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/prerenderHandler.js#L170) callback I am just returning `return console.log('CHECK THE GH ISSUE - https://github.com/redwoodjs/redwood/issues/6981#issuecomment-1357009132')` at the first line. **I'm not putting anything actionable in the Listr task.** 

Instead I've placed a `for of` loop  right [here](https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/prerenderHandler.js#L147) that iterates through the routes to be pre-rendered and runs everything that was previously defined in the Listr callback on line 170.

My hypothesis is that the crux of the issue is by throwing everything into a list of Listr tasks, the execution is happening concurrently in a `Promise.all` or similar. This creates the wrong Helmet context value globally and causes the pre-render to use those values. I also suspect this creates a pretty large memory footprint.

The concurrency is good for performance, but the React Helmet context needs to be scoped to each route pre-render instead of what appears to be globally.

You know this code way better than me, but my naive solutions are below. 

**Quick Fix?**: Make the pre-renders happen sequentially like my script above does.  Slower, but ensures correct context values for each route. 

**Proper Fix?**:  Scope the React Helmet context (and any other globals)  to each individual pre-render process, and batch the Listr tasks to reduce memory consumption. For example the message would look like `Now pre-rendering 50 BlogPosts 1/7` or something?

**Comment by dac09** [comment]
Thanks Elliot, that's very helpful. Your theory seems sound, and it could be related to our recent change from Listr to Listr2 (definitely something I tested before).

I'm finally back on my computer after a long break - I'll validate your hypothesis and atleast get the quickfix in! ✌️

**Comment by ched-dev** [comment]
A fix for this has been implemented in v4.0.0. Run a `yarn rw upgrade` to upgrade to latest version, and follow the [v4.0.0 Upgrade Guide](https://community.redwoodjs.com/t/redwood-v4-0-0-upgrade-guide/4412).
